nntrainer/engine.cpp

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file   engine.cpp
+ * @date   27 December 2024
+ * @brief  This file contains engine context related functions and classes that
+ * manages the engines (NPU, GPU, CPU) of the current environment
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <dirent.h>
+#include <dlfcn.h>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <app_context.h>
+#include <base_properties.h>
+#include <context.h>
+#include <engine.h>
+
+#ifdef ENABLE_OPENCL
+#include <cl_context.h>
+#endif
+
+static std::string solib_suffix = ".so";
+static std::string contextlib_suffix = "context.so";
+static const std::string func_tag = "[Engine] ";
+
+namespace nntrainer {
+
+std::mutex engine_mutex;
+
+std::once_flag global_engine_init_flag;
+
+void Engine::add_default_object(Engine &eg) {
+  /// @note all layers should be added to the app_context to guarantee that
+  /// createLayer/createOptimizer class is created
+
+  nntrainer::AppContext *app_context = new nntrainer::AppContext();
+  app_context->Global();
+
+  eg.registerContext("cpu", app_context);
```

Comment:
Could we swith registerContext to `std::unique_ptr`?
May `AppContext::Global()` throw now or in the future?

Or at least 
```c++
auto app_context = std::make_unique<AppContext>();
app_context->Global();
eg.regiestContext("cpu", app_context.release());
```
To make it exception-safety future-proof.